[TensorExpr] Introduce ExternalCall nodes to TE IR.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51475 [TensorExpr] Introduce ExternalCall nodes to TE IR.**

ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

Differential Revision: [D26179083](https://our.internmc.facebook.com/intern/diff/D26179083)

Commit:
[TensorExpr] Introduce ExternalCall nodes to TE IR.

ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

Differential Revision: [D26179083](https://our.internmc.facebook.com/intern/diff/D26179083)

[ghstack-poisoned]

asuhan reviewed (approved):
Some of the `clang-tidy` warnings look worth fixing, the rest looks good to me.

Commit:
Update on "[TensorExpr] Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

Differential Revision: [D26179083](https://our.internmc.facebook.com/intern/diff/D26179083)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

Differential Revision: [D26179083](https://our.internmc.facebook.com/intern/diff/D26179083)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

Differential Revision: [D26179083](https://our.internmc.facebook.com/intern/diff/D26179083)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

Differential Revision: [D26179083](https://our.internmc.facebook.com/intern/diff/D26179083)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

Differential Revision: [D26179083](https://our.internmc.facebook.com/intern/diff/D26179083)

[ghstack-poisoned]
